src/lib/reasoners/intervals_intf.ml

```diff
@@ -0,0 +1,735 @@
+      empty and an [Empty] kind is returned; otherwise, the resulting union is
+      [NonEmpty]. *)
+
+  (** {1 Inspection}
+
+      The following functions are used to inspect the global bounds of an
+      union of intervals. *)
+
+  val lower_bound : t -> value bound * explanation
+  (** [lower_bound u] returns a pair [lb, ex] of a global lower bound and an
+      explanation that justifies that this lower bound is valid. *)
+
+  val upper_bound : t -> value bound * explanation
+  (** [upper_bound u] returns a pair [lb, ex] of a global upper bound and an
+      explanation that justifies that this upper bound is valid. *)
+
+  val value_opt : t -> (value * explanation) option
+  (** If [u] is a singleton {b in the current context}, [value_opt u] returns a
+      pair [(v, ex)] where [v] is the value of [u] in the current context and
+      [ex] justifies [v] being both a lower and upper bound.
+
+      This is more efficient than checking if the values returned by
+      [lower_bound] and [upper_bound] are equal. *)
+
+  (** {1 Iteration}
+
+      The following functions are used to iterate over the maximal disjoint
+      intervals composing the union {b in the current context}.  *)
+
+  val to_seq : t -> bnd interval Seq.t
+  (** Convert an union of interval to a sequence of maximal disjoint and
+      non-adjacent intervals.
+
+      {b Warning}: The sequence of intervals is only valid in the current
+      context. *)
+
+  val iter : (bnd interval -> unit) -> t -> unit
+  (** [iter f u] calls [f] on each of the maximal disjoint and non-adjacent
+      intervals that make up the union [u] {b in the current context}. *)
+
+  val fold : ('a -> bnd interval -> 'a) -> 'a -> t -> 'a
+  (** [fold f acc u] folds [f] over each of the maximal disjoint and
+      non-adjacent intervals that make up the union [u] {b in the current
+      context}. *)
+
+  (** {1 Comparison}
+
+      The following functions are used to compare intervals {b in the current
+      context}. *)
+
+  val subset : ?strict:bool -> t -> t -> bool
+  (** [subset ?strict u1 u2] returns [true] if [u1] is a subset of [u2] {b in
+      the current context}. [subset] ignores all explanations.
+
+      If set, the [strict] flag ([false] by default) checks for strict
+      inclusion. *)
+
+  val equal : t -> t -> bool
+  (** [equal u1 u2] returns [true] if [u1] is equal to [u2] {b in the current
+      context}. [equal] ignores all explanations. *)
+
+  (** {1 Set manipulation} *)
+
+  val add_explanation : ex:explanation -> t -> t
+  (** [add_explanation ~ex u] adds the explanation [ex] to the union [u].
+
+      More precisely: [add_explanation ~ex u] is equivalent to [u] in models
+      where [ex] holds, and is [Interval.full] otherwise. *)
+
+  val intersect : t -> t -> (t, explanation) kind
+  (** [intersect u1 u2] computes the intersection of [u1] and [u2].
+
+      The resulting intersection is valid in all models. *)
```

Comment:
I'm not sure to understand what you mean by `valid in all models`?

Reply:
I mean that: in any model `M`, `intersect u1 u2` evaluates to the intersection of `M(u1)` and `M(u2)` (I guess it should rather be in all contexts). IOW; `intersect` is correct w.r.t explanations; I will clarify.

Reply:
> in any model M, `intersect u1 u2` evaluates to the intersection of `M(u1)` and `M(u2)`

Sorry, it should be `intersect u1 u2` evaluates to *a superset of* the intersection of `M(u1)` and `M(u2)` — in other contexts we might lose precision (but in the current context, it evaluates to the intersection exactly).